decryptor/mysql/data_encoder.go

```diff
@@ -0,0 +1,388 @@
+				return PutLengthEncodedString([]byte(*value)), true, nil
+			}
+			return data, false, ErrConvertToDataType
+		}
+		return data, false, nil
+	case common.EncryptedType_Bytes:
+		if !base.IsDecryptedFromContext(ctx) {
+			if newValue := setting.GetDefaultDataValue(); newValue != nil {
+				binValue, err := base64.StdEncoding.DecodeString(*newValue)
+				if err != nil {
+					return nil, false, err
+				}
+				return PutLengthEncodedString(binValue), true, nil
+			}
+			return data, false, ErrConvertToDataType
+		}
+		return data, false, nil
+	case common.EncryptedType_Int32, common.EncryptedType_Int64:
+		_, err := strconv.ParseInt(string(data), 10, 64)
+		// if it's valid string literal and decrypted, return as is
+		if err == nil {
+			return PutLengthEncodedString(data), true, nil
+		}
+		// if it's encrypted binary, then it is binary array that is invalid int literal
+		if !base.IsDecryptedFromContext(ctx) {
+			if newVal := setting.GetDefaultDataValue(); newVal != nil {
+				return PutLengthEncodedString([]byte(*newVal)), true, nil
+			}
+			return data, false, ErrConvertToDataType
+		}
+	}
+
+	return data, false, nil
+}
+
+func (p *BaseMySQLDataProcessor) encodeText(ctx context.Context, data []byte, setting config.ColumnEncryptionSetting, columnInfo base.ColumnInfo, logger *logrus.Entry) (context.Context, []byte, error) {
+	logger = logger.WithField("column", setting.ColumnName()).WithField("decrypted", base.IsDecryptedFromContext(ctx))
+	logger.Debugln("Encode text")
+	if len(data) == 0 {
+		// we still need to encode result data as it might be null field in db
+		return ctx, PutLengthEncodedString(data), nil
+	}
+
+	dataTypeEncoded, isEncoded, err := p.encodeTextWithDataType(ctx, data, setting)
+	if err != nil && err != ErrConvertToDataType {
+		return nil, nil, err
+	}
+
+	// in case of successful encoding with defined data type return encoded data
+	if isEncoded {
+		return ctx, dataTypeEncoded, nil
+	}
+
+	// in case of error on converting to defined type we should roll back field type and encode it as it was originally
+	if err == ErrConvertToDataType {
+		ctx = base.MarkErrorConvertedDataTypeContext(ctx)
+	}
+
+	return ctx, PutLengthEncodedString(data), nil
+}
+
+func (p *BaseMySQLDataProcessor) decodeBinary(ctx context.Context, encoded []byte, setting config.ColumnEncryptionSetting, columnInfo base.ColumnInfo, logger *logrus.Entry) (context.Context, []byte, error) {
+	// convert from binary to text literal because tokenizer expects int value as string literal
+	columnType := columnInfo.DataBinaryType()
+	if originType := columnInfo.OriginBinaryType(); originType != 0 {
+		columnType = originType
+	}
+
+	switch Type(columnType) {
+	case TypeNull:
+		// do nothing
+
+	case TypeTiny:
+		var numericValue int8
+		err := binary.Read(bytes.NewReader(encoded), binary.LittleEndian, &numericValue)
+		if err != nil {
+			break
+		}
+		return ctx, []byte(strconv.FormatInt(int64(numericValue), 10)), nil
```

Comment:
what about to use [strconv.AppendInt](https://pkg.go.dev/strconv#AppendInt)? It helps avoid extra conversion from `string` to `bytes`:
`strconv.FormatInt(nil, int64(numericValue), 10)`
same below

Reply:
Cool, didnt know about this methods before